### PR TITLE
Adjust mixed mode rules

### PR DESCRIPTION
**Pull request by eernstg**
Complementing https://dart-review.googlesource.com/c/sdk/+/139817, this PR proposes a few adjustments to the nnbd spec in locations where the description of mixed-mode inheritance differs from recent discussions about the approach and the implementation.

### CODE REVIEW

File: accepted/future-releases/nnbd/feature-specification.md [hunk 1]
```diff
@@ -1137,16 +1143,21 @@ arguments) marked as nullable or non-nullable as written.
 If a class `C` in an opted-in library implements the same generic class `I` more
 than once as `I0, .., In`, and at least one of the `Ii` is not syntactically
 equal to the others, then it is an error if `NNBD_TOP_MERGE(S0, ..., Sn)` is not
-defined where `Si` is **NORM(`Ii`)**.  Otherwise, for the purposes of runtime
-subtyping checks, `C` is considered to implement the canonical interface given
-by `NNBD_TOP_MERGE(S0, ..., Sn)`.
+defined where `Si` is **NORM(`Ii`)**.
+
+If `C` implements `I` once, or `C` implements `I` multiple times without error,
```
Comment (johnniwinther) [inline accepted/future-releases/nnbd/feature-specification.md]: This means that for
```Dart
  // opt-out
  class A implement Foo<FutureOr<Object>> {}
  // opt-in
  class B implements A {} 
```
`A` implements `Foo<FutureOr<Object>>` whereas `B` implements `Foo<Future<Object>>`. Is this intentional?

File: accepted/future-releases/nnbd/feature-specification.md [hunk 2]
```diff
@@ -992,9 +996,12 @@ arguments) marked as legacy types.
 
 If a class `C` in a legacy library implements the same generic class `I` more
 than once, it is an error if the `LEGACY_ERASURE` of all such super-interfaces
-are not all syntactically equal.  For the purposes of runtime subtyping checks,
-`C` is considered to implement the canonical `LEGACY_ERASURE` of the
-super-interfaces in question.
+are not all syntactically equal.
+
+When `C` implements `I` once, and also when `C` implements `I` more than once
+without error, `C` is considered to implement the canonical `LEGACY_ERASURE` of
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/nnbd/feature-specification.md]: I think my intention was to say something to the effect of "is considered to implement the canonical signature given by `LEGACY_ERASURE` of the super-interfaces in question".

File: accepted/future-releases/nnbd/feature-specification.md (hunk 1 above)
Comment (leafpetersen) [reply accepted/future-releases/nnbd/feature-specification.md]: I'm not keen on this change.  My goal here was to only invent a new type if required.  Why this change?

File: accepted/future-releases/nnbd/feature-specification.md [hunk 3]
```diff
@@ -992,9 +996,12 @@ arguments) marked as legacy types.
 
 If a class `C` in a legacy library implements the same generic class `I` more
 than once, it is an error if the `LEGACY_ERASURE` of all such super-interfaces
-are not all syntactically equal.  For the purposes of runtime subtyping checks,
-`C` is considered to implement the canonical `LEGACY_ERASURE` of the
-super-interfaces in question.
+are not all syntactically equal.
+
+When `C` implements `I` once, and also when `C` implements `I` more than once
```
Comment (leafpetersen) [inline accepted/future-releases/nnbd/feature-specification.md]: So just to be clear, I think this means that if a legacy class `B` has `A<int?>` as a super-interface, then in an opted in library, an assignment of the form `A<int> x = B()` is not considered an error?

File: accepted/future-releases/nnbd/feature-specification.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (eernstg) [reply accepted/future-releases/nnbd/feature-specification.md]: Using that wording.

File: accepted/future-releases/nnbd/feature-specification.md (hunk 1 above)
Comment (eernstg) [reply accepted/future-releases/nnbd/feature-specification.md]: For the example, normalization would reduce `Foo<FutureOr<Object*>*>` to `Foo<Object*>`, which eliminates `FutureOr`, and because `Object**` normalizes to `Object*`. So `A` implements `Foo<Object*>`.

And `B` implements `Foo<Object*>` because that's the `LEGACY_ERASURE` of `Foo<Object*>` (actually I think that erasure is `Foo<Object*>*`, but I don't think we want a rule that makes implemented types legacy at the top level, so we may need a variant of `LEGACY_ERASURE` that doesn't add a `*` at the top level, or maybe `LEGACY_ERASURE` just shouldn't add `*` at the top level at all).

But, @leafpetersen, I understand that you view this rule as particularly aggressive in its introduction of legacy typing. I used that wording in order to obtain a property that @johnniwinther mentioned a while back: A legacy class is purely legacy.

There may be different trade-offs here ("keep null-safe types whenever possible", in particular when `I` is only implemented once, versus "make sure that all properties of a legacy class are expressed in terms of legacy types").

@johnniwinther, WDYT?

File: accepted/future-releases/nnbd/feature-specification.md (hunk 3 above)
Comment (eernstg) [reply accepted/future-releases/nnbd/feature-specification.md]: Right.

This is a loss in static type safety. As I mentioned, it was motivated by the rule that a legacy class should only have legacy properties.

We can of course offer additional help to developers in their quest to detect discrepancies like this (for instance, a lint or a hint could flag such situations as dangerous).

However, `B` could change the actual member signatures and I do not believe that we'd have a rule justifying that `B` would then implement `I` in a different way:

```dart
// Opted in.
abstract class A<X> {
  X get x;
}

abstract class A1 extends A<int> {}

// Opted out.
class B implements A1 {
  int get x => null;
}

// Opted in.
class C implements B {}

void main() {
  A<int> a = C(); // OK!
}
```

So here we trust `C` to implement `A<int>`, even though the overriding implementation of `x` in `B` implies that `x` may actually return null (which means that it is reasonable to say that `C` implements `A<int*>`, but misleading to say that it implements `A<int>`).

So even though it looks like we are preserving more static type information, I suspect that it may not be sufficiently reliable to be very helpful.

File: accepted/future-releases/nnbd/feature-specification.md (hunk 1 above)
Comment (johnniwinther) [reply accepted/future-releases/nnbd/feature-specification.md]: For me difference is in whether we should perform normalization or not. If there is only one type we don't need to merge and therefore just use the type itself. In this case that'll make `B` implement `Foo<FutureOr<Object*>*>`. If there are more types we need to perform `NNBD_TOP_MERGE` and in this case we perform `NORM` on the input:
```Dart
// opt-out
class A implements Foo<FutureOr<Object>> {}
// opt-in
class B implements Foo<FutureOr<dynamic>> {}
class C implements A, B {}
```
Here `C` implements 
```
NNBD_TOP_MERGE(NORM(Foo<FutureOr<Object*>*>),NORM(Foo<FutureOr<dynamic>>))
  = NNBD_TOP_MERGE(Foo<Object*>,Foo<dynamic>)
  = Foo<Object?>
```

Note that I don't consider the outmost declaration in extends, implements and with clauses to have nullability so `Foo` is just `Foo` regardless of opt-in/out status.

Comment (leafpetersen) [reply accepted/future-releases/nnbd/feature-specification.md]: @eernstg I don't understand your answer.  The issue that @johnniwinther was pointing out and that I was responding to has nothing to do with legacy, as far as I can tell.  My original spec only applied NNBD_TOP_MERGE when there were multiple super-interfaces with the same generic base class.  You add the the line `If `C` implements `I` once,`.  This line is the one that I question.  It implies we normalize all super-interfaces, which is a much higher impact change.

Comment (eernstg) [reply accepted/future-releases/nnbd/feature-specification.md]: The "when `C` implements `I` once" change should only be applied when `C` is legacy, it was a mistake to have it here. Sorry about that.

Comment (eernstg) [reply accepted/future-releases/nnbd/feature-specification.md]: However, I would expect an implementation to be allowed to normalize _any_ term derived from `<type>` as eagerly as it wishes. Some of these terms may have syntax like `FutureOr<X>` where `X` is a type variable, and this may obviously call for normalization at run time, but with `FutureOr<Object?>` I would expect it to be OK for a compiler to normalize it at compile-time.

@johnniwinther, do you prefer to avoid compile-time type normalization except in the case where it is used to resolve a conflict among implemented superinterfaces or inherited member signatures? I guess the point would be that normalization which is eager to some unknown/unspecified degree would create problems in the management of the values of constant expressions? (Some constant expressions suddenly evaluate to the same object, even though they look different).

I think that issue might not arise with the normalization that is applied during the two cases of conflict resolution that are in focus here, because canonicalization would only ever identify instances of the same class, and if we have two instances of the given class `C` then it makes no difference that they implement a normalized version of some superinterface, because that will be true for both of them.

**Comment by leafpetersen** [comment]
Once we settle on the tests, can you file issues for the analyzer and CFE to implement this change?

Also, consider adding an entry to the CHANGELOG at the top?

File: accepted/future-releases/nnbd/feature-specification.md (hunk 1 above)
Comment (johnniwinther) [reply accepted/future-releases/nnbd/feature-specification.md]: I prefer to only normalize when we need to synthesize a result (e.g. in case of implementing to instantiations of a generic interface).

Comment (eernstg) [reply accepted/future-releases/nnbd/feature-specification.md]: I think there are no further issues here: Normalizing when a synthesized result is needed is necessary, and if we wish to introduce normalization elsewhere then it would be a non-breaking change. So I'll resolve this thread.

**Comment by eernstg** [comment]
Thanks! I updated the CHANGELOG entry (this is line 9 of feature-specification.md, I assume that's what we are talking about). I'll file analyzer/cfe implementation issues when the [test CL](https://dart-review.googlesource.com/c/sdk/+/139817) lands. I see that it is fully LGTM'd, so I'll go ahead and pre-approve the test failures and land it.

@scheglov, you haven't indicated your approval of this PR yet; if you do have further comments, please speak up!

**Comment by scheglov** [comment]
@eernstg LGTM

**Comment by eernstg** [comment]
Thanks!
